intellij/src/saros/core/ui/eventhandler/XMPPAuthorizationHandler.java

```diff
@@ -1,46 +1,48 @@
 package saros.core.ui.eventhandler;
 
 import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
 import java.text.MessageFormat;
-import org.apache.log4j.Logger;
 import saros.intellij.ui.Messages;
 import saros.intellij.ui.util.DialogUtils;
+import saros.intellij.ui.util.UIProjectUtils;
 import saros.net.xmpp.JID;
 import saros.net.xmpp.subscription.SubscriptionHandler;
 import saros.net.xmpp.subscription.SubscriptionListener;
 
 /** Handler for accepting or rejecting incoming XMPP subscription requests */
 public class XMPPAuthorizationHandler {
+  private final UIProjectUtils projectUtils;
 
-  private static final Logger LOG = Logger.getLogger(XMPPAuthorizationHandler.class);
+  @SuppressWarnings("FieldCanBeLocal")
   private final SubscriptionListener subscriptionListener =
       new SubscriptionListener() {
 
         @Override
         public void subscriptionRequestReceived(final JID jid) {
 
-          ApplicationManager.getApplication()
-              .invokeLater(
-                  new Runnable() {
-                    @Override
-                    public void run() {
-                      handleAuthorizationRequest(jid);
-                    }
-                  });
+          projectUtils.runWithProject(
+              project ->
+                  ApplicationManager.getApplication()
+                      .invokeLater(() -> handleAuthorizationRequest(project, jid)));
         }
       };
+
   private final SubscriptionHandler subscriptionHandler;
 
-  public XMPPAuthorizationHandler(final SubscriptionHandler subscriptionHandler) {
+  public XMPPAuthorizationHandler(
+      final SubscriptionHandler subscriptionHandler, UIProjectUtils projectUtils) {
     this.subscriptionHandler = subscriptionHandler;
     this.subscriptionHandler.addSubscriptionListener(subscriptionListener);
+
+    this.projectUtils = projectUtils;
```

Comment:
Even if its actual potential should be low, this could lead to a race condition:
If the `subscriptionListener` is triggered while `projectUtils` is not initialized / visible, it will fail. Reordering the statements won't help.

From the [Java Language Specification](
https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5):
>An object is considered to be completely initialized when its constructor finishes. A thread that can only see a reference to an object after that object has been completely initialized is guaranteed to see the correctly initialized values for that object's final fields.


A Solution would be to make the `anonymous class` a `static final` one and move all references into it. Or register the Listener outside the constructor.

Reply:
You could also make the variable volatile and assign it first. But my guess is that this class is a singleton used inside the PicoContainer ? and so this variable is always initialized before the listener is triggered.

Reply:
Yes, the XMPPAuthorizationHandler is instantiated by the PicoContainer plugin context. Furthermore, the handler reacts to incoming XMPP Subscription events, which require a running connection to an XMPP account.

I could let the class implement `Startable` and move the subscription to the `start()` call. But I am not sure whether this is really necessary. What are your opinions?

I have seen (and used) this pattern multiple times before, so if it is an issue, we should also aim to adjust all other places where listeners with dependencies on other instance variables are subscribed in the CTOR.

Reply:
The main container is not started at all because in former times some components reinjected themselves and so it was prohibited by PicoContainer to call start on such a container. I guess that can be changed now. But I do not know if it is worth the effort. And yes this pattern is used multiples times all over the code (registering listeners inside the CTOR).

Reply:
> The main container is not started at all because in former times some components reinjected themselves and so it was prohibited by PicoContainer to call start on such a container.

Good to know. I will have to make sure that I don't rely on Startable in such cases then.

Will the PicoContainer just 'silently' not run the start/stop methods or will it complain about Startable components in the main container?

Reply:
IIRC the reinject fails if the container was already started. However we have no components now that want to reinject themselves. I think you have to play around a little bit and make sure that https://github.com/saros-project/saros/blob/master/core/src/saros/context/ContainerContext.java#L250 is still working if you called start on the main container.

Reply:
So this is not optimal, but as long as no other thread will call the listener during the startup (like a ui thread or something) it's ok.

> You could also make the variable volatile and assign it first. 

This is by Specification not a safe approach, because the anonymous class still gets a `this` reference to the incomplete object.

Reply:
Looking at the code I have my doubts that all final fields would not be written at the time you call addSubscriptionListener (assuming that this would be the last statement in the CTOR). But yes it would be only confusing to declare each variable as volatile to ensure that the callback would not see a half initialized class.

Reply:
I think I will go ahead an merge this PR as the issue is not pressing in this case and a more general solution approach can be discussed in #534.